tempodb/encoding/vparquet5/row_number_iterator.go

```diff
@@ -0,0 +1,141 @@
+package vparquet5
+
+import (
+	"errors"
+
+	pq "github.com/grafana/tempo/pkg/parquetquery"
+)
+
+type rowNumberIterator struct {
+	rowNumbers []pq.RowNumber
+}
+
+var _ pq.Iterator = (*rowNumberIterator)(nil)
+
+func (r *rowNumberIterator) String() string {
+	return "rowNumberIterator()"
+}
+
+func (r *rowNumberIterator) Next() (*pq.IteratorResult, error) {
+	if len(r.rowNumbers) == 0 {
+		return nil, nil
+	}
+
+	res := &pq.IteratorResult{RowNumber: r.rowNumbers[0]}
+	r.rowNumbers = r.rowNumbers[1:]
+	return res, nil
+}
+
+func (r *rowNumberIterator) SeekTo(to pq.RowNumber, definitionLevel int) (*pq.IteratorResult, error) {
+	var at *pq.IteratorResult
+
+	for at, _ = r.Next(); r != nil && at != nil && pq.CompareRowNumbers(definitionLevel, at.RowNumber, to) < 0; {
+		at, _ = r.Next()
+	}
+
+	return at, nil
+}
+
+func (r *rowNumberIterator) Close() {}
+
+var _ pq.Iterator = (*virtualRowNumberIterator)(nil)
+
+// virtualRowNumberIterator is an iterator that reads the row count from a column in a parquet file and uses this count
+// to calculate virtual row numbers on the given definitionLevel. This is used to iterate over spans in a trace without
+// actually reading a column on span definition level.
+type virtualRowNumberIterator struct {
+	iter            pq.Iterator // iterator returning row count
+	definitionLevel int
+
+	at       pq.IteratorResult
+	rowsMax  int32
+	rowsLeft int32
+}
+
+func newVirtualRowNumberIterator(iter pq.Iterator, definitionLevel int) *virtualRowNumberIterator {
+	return &virtualRowNumberIterator{
+		iter:            iter,
+		definitionLevel: definitionLevel,
+		at:              pq.IteratorResult{RowNumber: pq.EmptyRowNumber()},
+	}
+}
+
+func (v *virtualRowNumberIterator) String() string {
+	return "virtualRowNumberIterator()"
+}
+
+func (v *virtualRowNumberIterator) Next() (*pq.IteratorResult, error) {
+	if v.rowsLeft == 0 {
+		res, err := v.iter.Next()
+		if err != nil {
+			return nil, err
+		}
+		if res == nil {
+			return nil, nil
+		}
+		if len(res.Entries) != 1 {
+			return nil, errors.New("expecting exactly one entry")
+		}
+
+		v.at.RowNumber = res.RowNumber
+		v.setRows(res.Entries[0].Value.Int32())
+
+		if v.rowsLeft == 0 {
+			return &v.at, nil
+		}
+	}
+
+	v.rowsLeft--
+	v.at.RowNumber.Next(v.definitionLevel, v.definitionLevel, v.definitionLevel)
+
+	return &v.at, nil
```

Comment:
this is returning a pointer to the result but holding it for future use which might be unique? i don't necessarily know of any place in the fetch layer that we modify an iterator result but it's potentially an issue.

going to dig through the code to see if i find a place an iterator results is modified.